File: Sorted Container/SortedDict/2213. Longest Substring of One Repeating Character/solution.py
from sortedcontainers import SortedList, SortedDict
class Solution:
    def longestRepeating(self, s: str, queryCharacters: str, queryIndices: List[int]) -> List[int]:
        n, k = len(s), len(queryCharacters)

        interval = SortedDict() # {start_index: length}
        curIdx = -1
        for i in range(n):
            if i == 0 or s[i] != s[i-1]:
                interval[i] = 1
                curIdx = interval.index(i)
            else:
                k, v = interval.peekitem(curIdx)
                interval[k] += 1

        length = SortedList()
        for l in interval.values():
            length.add(l)
   
        arr = list(s)
        res = []
        for i, ch in zip(queryIndices, queryCharacters):
            if ch != arr[i]:
                arr[i] = ch
                idx = interval.bisect_right(i)-1
                
                # break interval
                j, size = interval.peekitem(idx)
                # j______i
                # i j_____
                # j__ i j_
                if size > 1:
                    del interval[j]
                    length.remove(size)

                    if i == j:
                        interval[i] = 1
                        length.add(1)

                        interval[j+1] = size-1
                        length.add(size-1)
                        
                    elif i == j+size-1:
                        interval[i] = 1
                        length.add(1)

                        interval[j] = size-1
                        length.add(size-1)
                    else:
                        interval[i] = 1
                        length.add(1)
                        
                        interval[j] = i-j
                        length.add(interval[j])

                        interval[i+1] = size - (i-j+1)
                        length.add(interval[i+1])

                # merge interval
                ## check merge right
                if i+1<n and ch == arr[i+1]:
                    idx = interval.bisect_right(i)
                    j, size = interval.peekitem(idx)
                    length.remove(size)
                    length.remove(interval[i])
                    del interval[j]

                    interval[i] = interval[i]+size
                    length.add(interval[i])
                
                ## check merge left
                if i-1 >= 0 and ch == arr[i-1]:
                    l = interval.bisect_left(i)-1
                    j, size = interval.peekitem(l)
                    length.remove(interval[i])
                    length.remove(size)
                    interval[j] = size+interval[i]
                    length.add(interval[j])
                    del interval[i]
            res.append(length[-1])
        return res

File: Sorted Container/SortedDict/2213. Longest Substring of One Repeating Character/solution.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
private:
    struct Node {
        int max_len;
        int prefix_len;
        char prefix_char;
        int suffix_len;
        char suffix_char;
        int length;
    };

    vector<Node> tree;

    Node merge(const Node& left, const Node& right) {
        Node res;
        res.length = left.length + right.length;

        // Default prefix initialization
        res.prefix_char = left.prefix_char;
        res.prefix_len = left.prefix_len;
        if (left.prefix_len == left.length && left.suffix_char == right.prefix_char) {
            res.prefix_len = left.length + right.prefix_len;
        }

        // Default suffix initialization
        res.suffix_char = right.suffix_char;
        res.suffix_len = right.suffix_len;
        if (right.suffix_len == right.length && right.prefix_char == left.suffix_char) {
            res.suffix_len = right.length + left.suffix_len;
        }

        // Calculate max length
        res.max_len = max(left.max_len, right.max_len);
        if (left.suffix_char == right.prefix_char) {
            res.max_len = max(res.max_len, left.suffix_len + right.prefix_len);
        }

        return res;
    }

    void build(int node, int start, int end, const string& s) {
        if (start == end) {
            tree[node] = {1, 1, s[start], 1, s[start], 1};
            return;
        }
        int mid = start + (end - start) / 2;
        build(2 * node, start, mid, s);
        build(2 * node + 1, mid + 1, end, s);
        tree[node] = merge(tree[2 * node], tree[2 * node + 1]);
    }

    void update(int node, int start, int end, int idx, char ch) {
        if (start == end) {
            tree[node] = {1, 1, ch, 1, ch, 1};
            return;
        }
        int mid = start + (end - start) / 2;
        if (idx <= mid) {
            update(2 * node, start, mid, idx, ch);
        } else {
            update(2 * node + 1, mid + 1, end, idx, ch);
        }
        tree[node] = merge(tree[2 * node], tree[2 * node + 1]);
    }

public:
    vector<int> longestRepeating(string s, string queryCharacters, vector<int>& queryIndices) {
        int n = s.length();
        int k = queryIndices.size();
        tree.resize(4 * n);

        build(1, 0, n - 1, s);

        vector<int> ans(k);
        for (int i = 0; i < k; ++i) {
            update(1, 0, n - 1, queryIndices[i], queryCharacters[i]);
            ans[i] = tree[1].max_len;
        }

        return ans;
    }
};